lldb/source/Core/Telemetry.cpp

```diff
@@ -0,0 +1,619 @@
+}
+
+void BasicTelemeter::LogCommandEnd(llvm::StringRef uuid,
+                                   llvm::StringRef command_name,
+                                   llvm::StringRef command_args,
+                                   TelemetryEventStats stats,
+                                   Target *target_ptr,
+                                   CommandReturnObject *result) {
+
+  lldb_private::CommandTelemetryInfo command_info =
+      MakeBaseEntry<lldb_private::CommandTelemetryInfo>();
+
+  // If we have a target attached to this command, then get the UUID.
+  command_info.target_uuid = "";
+  if (target_ptr && target_ptr->GetExecutableModule() != nullptr) {
+    command_info.target_uuid =
+        target_ptr->GetExecutableModule()->GetUUID().GetAsString();
+  }
+  command_info.command_uuid = uuid.str();
+  command_info.command_name = command_name.str();
+  command_info.args = command_args.str();
+  command_info.stats = std::move(stats);
+  command_info.exit_description = {result->Succeeded() ? 0 : -1,
+                                   StatusToString(result)};
+  EmitToDestinations(&command_info);
+}
+
+llvm::StringRef parse_value(llvm::StringRef str, llvm::StringRef label) {
+  return str.substr(label.size()).trim();
+}
+
+bool parse_field(llvm::StringRef str, llvm::StringRef label) {
+  if (parse_value(str, label) == "true")
+    return true;
+  return false;
+}
+
+<<<<<<< HEAD
+llvm::telemetry::TelemetryConfig *MakeTelemetryConfig() {
+  bool enable_telemetry = false;
+  std::vector<std::string> additional_destinations;
+
+  // Look in the $HOME/.lldb_telemetry_config file to populate the struct
+  llvm::SmallString<64> init_file;
+  FileSystem::Instance().GetHomeDirectory(init_file);
+  llvm::sys::path::append(init_file, ".lldb_telemetry_config");
+  FileSystem::Instance().Resolve(init_file);
+  if (llvm::sys::fs::exists(init_file)) {
+    auto contents = llvm::MemoryBuffer::getFile(init_file, /*IsText*/ true);
+    if (contents) {
+      llvm::line_iterator iter =
+          llvm::line_iterator(contents->get()->getMemBufferRef());
+      for (; !iter.is_at_eof(); ++iter) {
+        if (iter->starts_with("enable_telemetry:")) {
+          enable_telemetry = parse_field(*iter, "enable_telemetry:");
+        } else if (iter->starts_with("destination:")) {
+          llvm::StringRef dest = parse_value(*iter, "destination:");
+          if (dest == "stdout") {
+            additional_destinations.push_back("stdout");
+          } else if (dest == "stderr") {
+            additional_destinations.push_back("stderr");
+          } else {
+            additional_destinations.push_back(dest.str());
+          }
+        }
+      }
+    } else {
+      std::cerr << "Error reading config file at " << init_file.c_str() << "\n";
+    }
+  }
+
+  auto *ret = new llvm::telemetry::TelemetryConfig{enable_telemetry,
+                                                   additional_destinations};
+#ifdef HAS_VENDOR_TELEMETRY_CONFIG
```

Comment:
can you clarify what you mean by dummy implementations? (w/o causing either duplicate-symbols errors) I'd tried the weak-def but it wasn't very portable, so I thought this was the simplest.

Reply:
I just meant that we could conditionally provide an empty implementation of the "vendor" function instead of conditionally calling it. So something like:
```
#if VENDOR
// do something to get the vendor definition
#else
// provide an empty definition ourselves, e.g:
void ApplyVendorSpecificConfigs(...) {}
#endif
```

Then you don't need the ifdef at the call site.

Reply:
Done - reworked this to use Plugin registry (the default plugin is enabled for test, then vendors can add their own impl as needed)